### PR TITLE
[FEATURE] FnShutdown service with on_shutdown callback

### PR DESCRIPTION
**Pull request by leon3s**
Hey following our conversation i tried to implement the `FnShutdown` Service.
Maybe i should have used `FnServiceFactory` instead of `Service` ?

### CODE REVIEW

**Comment by fafhrd91** [comment]
fn_shutdown should provide only shutdow callback and nothing else. usage would be 

```rust
pipeline(srv).and_then(fn_shutdown(|| some code))
```

**Comment by leon3s** [comment]
I see, sorry i'm not 100% familiar with all your API, i'm gonna change that!

**Comment by leon3s** [comment]
I tried to prototype it the way you wanted in previous comments

```rust
FnShutdown<S, F: FnOnce>
```

But i had to implement the `call` method of the `Service` trait and keep traking the Error.
Im not sure about the :
```rust
Req: Clone + 'static 
```
aswell but i didn't find other way to make it compile..

**Comment by fafhrd91** [comment]
here is working code, I think it would be better to wrap other service and make FnShutdown<S, F>{..}
in that case call method would actually call inner service, otherwise it just return result back

```rust

use std::{cell::Cell, future::ready, future::Ready, marker, task::Context, task::Poll};

use crate::Service;

pub struct FnShutdown<Req, Err, F = fn()> {
    f_shutdown: Cell<Option<F>>,
    _t: marker::PhantomData<(Req, Err)>,
}

impl<Req, Err, F> Service<Req> for FnShutdown<Req, Err, F>
where
    F: FnOnce(),
{
    type Response = Req;
    type Error = Err;
    type Future<'f> = Ready<Result<Req, Err>> where Self: 'f, Req: 'f;

    #[inline]
    fn poll_shutdown(&self, _: &mut Context<'_>) -> Poll<()> {
        if let Some(f) = self.f_shutdown.take() {
            (f)()
        }
        Poll::Ready(())
    }

    fn call(&self, req: Req) -> Self::Future<'_> {
        ready(Ok(req))
    }
}
```

**Comment by leon3s** [comment]
Got it thank you

File: ntex-service/src/fn_shutdown.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
+use std::{
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:1]: could you split into separate lines, I don't like such formatting

File: ntex-service/src/fn_shutdown.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    cell::Cell,
+    future::{ready, Ready},
+    marker::PhantomData,
+    task::{Context, Poll},
+};
+
+use crate::Service;
+
+#[inline]
+/// Create `FnShutdown` service with a on_shutdown callback.
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:11]: could you add detail doc, similar to fn_service

File: ntex-service/src/fn_shutdown.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+{
+    FnShutdown::new(f)
+}
+
+pub struct FnShutdown<Req, Err, F = fn()> {
+    f_shutdown: Cell<Option<F>>,
+    _t: PhantomData<(Req, Err)>,
+}
+
+impl<Req, Err, F> FnShutdown<Req, Err, F> {
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:24]: could you also add Clone implementation if F: Clone

File: ntex-service/src/lib.rs [hunk 4]
```diff
@@ -22,6 +23,7 @@ mod then;
 
 pub use self::apply::{apply_fn, apply_fn_factory};
 pub use self::fn_service::{fn_factory, fn_factory_with_config, fn_service};
+pub use self::fn_shutdown::fn_shutdown;
```
Comment (fafhrd91) [inline ntex-service/src/lib.rs:26]: also import FnShutdown in dev module

**Review by fafhrd91 — CHANGES_REQUESTED** [review]
could you also add entry to CHANGES, and bump version to 1.0.1

lgtm.

File: ntex-service/src/fn_shutdown.rs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        Self {
+            f_shutdown: Cell::new(Some(f)),
+            _t: PhantomData,
+        }
+    }
+}
+
+impl<Req, Err, F> Clone for FnShutdown<Req, Err, F>
+where
+    F: FnOnce(),
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:33]: this should be `F: Clone`

File: ntex-service/src/fn_shutdown.rs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+{
+    #[inline]
+    fn clone(&self) -> Self {
+        Self {
+            f_shutdown: Cell::new(self.f_shutdown.take()),
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:38]: this moves function, you need to clone it

File: ntex-service/src/fn_shutdown.rs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+#[inline]
+/// Create `FnShutdown` for function that can act as a `on_shutdown` callback.
+pub fn fn_shutdown<Req, Err, F>(f: F) -> FnShutdown<Req, Err, F>
+where
+    F: FnOnce(),
+{
+    FnShutdown::new(f)
+}
+
+pub struct FnShutdown<Req, Err, F = fn()> {
```
Comment (fafhrd91) [inline ntex-service/src/fn_shutdown.rs:17]: remove default `= fn()` arg, FnShutdown won't have default F ever

**Comment by leon3s** [comment]
Something like that ?

**Review by fafhrd91 — APPROVED** [review]
lgtm

**Comment by fafhrd91** [comment]
good work!

**Pull request merged**